### ISSUE TITLE
cassandra_ycsb AI result reporting is incorrect

### ISSUE THREAD

**Issue by cpschult** [ibmcb/cbtool#58]
It appears lazy_collection in scripts/cassandra_ycsb/cb_ycsb_common.sh may have a couple problems.

1)  lazy_collection (the default collection mechanism) differs from eager_collection in that it uses different units for latency measurements.  I'm having trouble capturing the output from the YCSB run command to verify the measurement units but I assume they are the same as used in the generation phase ( which are microseconds for average measurements ).
```
...
[OVERALL], RunTime(ms), 646267.0
[OVERALL], Throughput(ops/sec), 1547.3480774973812
[INSERT], Operations, 1000000
**[INSERT], AverageLatency(us), 641.728206**
[INSERT], MinLatency(us), 252
[INSERT], MaxLatency(us), 129799
[INSERT], 95thPercentileLatency(ms), 0
[INSERT], 99thPercentileLatency(ms), 1
[INSERT], Return=0, 1000000
...
```

I think the fix is pretty simple.  Something like this:
```
$diff a/scripts/cassandra_ycsb/cb_ycsb_common.sh b/scripts/cassandra_ycsb/cb_ycsb_common.sh
-    latency:$(expr $latency):ms \
+    latency:$(expr $latency):us \
```

2)  But there appears to be another problem.  It looks like lazy_collection looks for READ and UPDATE measurements in the output file.  (this seems OK for YCSB! workload 'workloadd' ) As written lazy_collection only reports one of the values and doesn't provide any indication which value is returned.  Should it be written similar to eager_collection where it reports both read and update values back to cb orchestrator?

I haven't tested the proposed changed below.  It seems there could be farther reaching consequences as 'latency' is no longer returned in the result set.  I don't know if that is a problem.

something like:
```
diff --git a/scripts/cassandra_ycsb/cb_ycsb_common.sh b/scripts/cassandra_ycsb/cb_ycsb_common.sh
index 00db685..46eeb5f 100755
--- a/scripts/cassandra_ycsb/cb_ycsb_common.sh
+++ b/scripts/cassandra_ycsb/cb_ycsb_common.sh
 

@@ -170,20 +174,20 @@ function lazy_collection {
         then
             if [[ ${array[1]} == *AverageLatency* ]]
             then
-                latency=${array[2]}
+                update_avg_latency=${array[2]}
             fi
         fi
         if [[ ${array[0]} == *READ* ]]
         then
             if [[ ${array[1]} == *AverageLatency* ]]
             then
-                latency=${array[2]}
+                read_avg_latency=${array[2]}
             fi
         fi
     
@@ -221,7 +225,8 @@ function lazy_collection {
     load_profile:${LOAD_PROFILE}:name \
     load_duration:${LOAD_DURATION}:sec \
     throughput:$(expr $ops):tps \
-    latency:$(expr $latency):us \
+    update_avg_latency:$(expr $update_avg_latency):us \
+    read_avg_latency:$(expr $read_avg_latency):us \
     datagen_time:${datagentime}:sec \
     datagen_size:${datagensize}:records \
     ${SLA_RUNTIME_TARGETS}

```

**Comment by cpschult** [comment]
Earlier I said:
>I'm having trouble capturing the output from the YCSB run command to verify the measurement units but I assume they are the same as used in the generation phase ( which are microseconds for average measurements ).

I had added a third parameter to lazy_collection which contained a temporary file name.  However the second parameter was empty and so the third parameter was always empty.  I re-ordered the parameters and now can capture the output from the YCSB command.

As expected, average latency measurements are always generated with microsecond units.  Therefore the units reported to cloudbench should be 'us' instead of 'ms' as suggested in the OP.

Also on the location of UPDATE and READ results in the file I think the latency value reported to cloudbench will always be the average  READ latency.

I made changes in two files to allow printing output from the RUN phase of YCSB. I think this is very helpful and should also be considered for addition to cloudbench.  ( Output from run phase is written to file /tmp/tmp.XXXXXXXX.run where XXXXXXXX is equal to the unique string from the LOAD phase of YCSB. )
```
diff --git a/scripts/cassandra_ycsb/cb_ycsb.sh b/scripts/cassandra_ycsb/cb_ycsb.sh
index b9dac9e..2ca53d4 100755
--- a/scripts/cassandra_ycsb/cb_ycsb.sh
+++ b/scripts/cassandra_ycsb/cb_ycsb.sh
@@ -75,6 +75,7 @@ sudo sh -c "echo "operationcount=$OPERATION_COUNT" >> $YCSB_PATH/custom_workload

 source ~/cb_barrier.sh start

+OUTPUT_FILE=$(mktemp)
 if [[ ${GENERATE_DATA} == "true" ]]
 then  
        FIRST_SEED=$(echo $seed_ips_csv | cut -d ',' -f 1)
@@ -85,7 +86,6 @@ then
        cassandra-cli -h ${FIRST_SEED} -f create_keyspace.cassandra

     syslog_netcat "Number of records to be inserted : $RECORDS"
-    OUTPUT_FILE=$(mktemp)

     log_output_command=$(get_my_ai_attribute log_output_command)
     log_output_command=$(echo ${log_output_command} | tr '[:upper:]' '[:lower:]')
@@ -129,13 +129,13 @@ then
     syslog_netcat "Command line is: ${CMDLINE}. Output file is ${OUTPUT_FILE}"
     if [[ $APP_COLLECTION == "lazy" ]]
     then
-        lazy_collection "$CMDLINE" ${SLA_RUNTIME_TARGETS}
+        lazy_collection "$CMDLINE" ${OUTPUT_FILE} ${SLA_RUNTIME_TARGETS} 
     else
        eager_collection "$CMDLINE" ${SLA_RUNTIME_TARGETS}
     fi
 else
     syslog_netcat "This AI reached the limit of load generation process executions. If you want this AI to continue to execute the load generator, reset the \"run_limit\" counter"
     sleep ${LOAD_DURATION}
 fi


diff --git a/scripts/cassandra_ycsb/cb_ycsb_common.sh b/scripts/cassandra_ycsb/cb_ycsb_common.sh
index 00db685..084d836 100755
--- a/scripts/cassandra_ycsb/cb_ycsb_common.sh
+++ b/scripts/cassandra_ycsb/cb_ycsb_common.sh
@@ -54,6 +54,7 @@ BACKEND_TYPE=$(get_my_ai_attribute type | sed 's/_ycsb//g')

 if [[ $BACKEND_TYPE == "cassandra" ]]
 then
+    export CASSANDRA_CONFIG_DIR=/etc/cassandra/
     CASSANDRA_DATA_DIR=$(get_my_ai_attribute_with_default cassandra_data_dir /dbstore)
     eval CASSANDRA_DATA_DIR=${CASSANDRA_DATA_DIR}

@@ -151,13 +152,15 @@ fi
 function lazy_collection {

     CMDLINE=$1
-    SLA_RUNTIME_TARGETS=$2
+    OUTPUT_FILE=$2.run
+    SLA_RUNTIME_TARGETS=$3

     ops=0
     latency=0

     while read line
     do
+        echo $line >> $OUTPUT_FILE
         IFS=',' read -a array <<< "$line"
         if [[ ${array[0]} == *OVERALL* ]]
         then
```

**Comment by maugustosilva** [comment]
Fixed in commit b089cfd4bd883824636498058b096842c87f6ca6. Thanks Chris.
